### PR TITLE
{Core} Add credential helper for cli

### PR DESCRIPTION
**Pull request by evelyn-ys**
**Related command**

**Description**

CLI commands may expose credentials in output and it may bring security risks. We need to design carefully how we should handle output credentials.

This PR
- adds `CredentialHelper` to automatically distinguish credentials from command output based on regexes
- For phase 1, we only show warnings to customers if customers proactively turn on this feature with `az config set clients.show_secrets_warning=True`
- also adds sensitive tag framework to allow manually tagging sensitive commands 

**Testing Guide**

First of all, turn on the feature
`az config set clients.show_secrets_warning=True`
Then run any commands that may expose credentials, eg.
`az storage account keys list --account-name yssa -g yishitest`
image

**History Notes**

[Core] Warn customers when there're credentials in cli output if this feature has been turned on through `az config set clients.show_secrets_warning=True`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli-core/azure/cli/core/credential_helper.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+class CredentialType(Enum):
+    STORAGE_SAS_TOKEN = ('([?&;])sig=[\w%-]+', r'\1sig=_REDACTED_SAS_TOKEN_SIG_', 'Storage SAS token')
+    CODE = ('([?&;])code=[\w%+/=-]+', r'\1code=_REDACTED_CODE_', 'Several types of keys/secrets are passed with a query parameter "code"')
+    AZCOPY_TOKEN = ('([?&/]|%2F)(SourceKey|DestKey)(=|%3A|%3D|:)(?:[\w/-]|%2F)+(?:==|%3D%3D)', r'\1\2\3_REDACTED_AZCOPY_TOKEN_', 'AzCopy token')
+    JWT_TOKEN = ('(?:eyJ0eXAi|eyJhbGci)[\w\-.~+/%]*', '_REDACTED_JWT_TOKEN_', 'JWT token')
+    BEARER_TOKEN = ('("|\'|%22)(bearer |bearer%20)[\w\-.~+/]{100,}("|\'|%22)', r'\1\2_REDACTED_BEARER_TOKEN_\3', 'Bearer token')
+    SSH_KEY = ('("|\')(ssh-rsa )AAAA[\w\-.~+/]{100,}("|\')', r'\1\2_REDACTED_SSH_KEY_\3', 'SSH key')
+    EMAIL_ADDRESS = ('\b[\w.%#+-]+(%40|@)([a-z0-9.-]*\.[a-z]{2,})\b', r'_REDACTED_EMAIL_\1\2', 'Email address')
+    GUID = ('([0-9a-f]{8}-?[0-9a-f]{4}-?[0-9a-f]{4}-?[0-9a-f]{4}-?[0-9a-f]{12})\b', '_REDACTED_GUID_', 'GUID')
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/credential_helper.py:19]: It is a good practice to always provide explanations and examples for regular expressions, as regular expressions can be difficult to understand for other code readers.

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/credential_helper.py]: Totally agree so I add test cases for each type

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/credential_helper.py]: also added comments as explanation

File: src/azure-cli-core/azure/cli/core/commands/sensitive.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    except KeyError:
+        command_group = _get_command_group(name)
+        group_kwargs = getattr(command_group, 'group_kwargs', None)
+        if group_kwargs:
+            sensitive_info = group_kwargs.get(_sensitive_kwarg, None)
+    return sensitive_info
+
+
+# pylint: disable=too-many-instance-attributes
+class SensitiveItem(StatusTag):
```
Comment (evelyn-ys) [inline src/azure-cli-core/azure/cli/core/commands/sensitive.py:34]: I still keep this new tag even though there's no command adding sensitive tag for now.

With this new tag, we have the ability to manually mark some properties as sensitive for specific command. For example, `az keyvault secret set` command will return `value` property which should be sensitive but may not be recognized by regexes.

**Review by calvinhzy — APPROVED** [review]

**Review by bebound — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Review by yonzhan — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/credential_helper.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+sensitive_data_warning_message = '[Warning] This output may compromise security by showing secrets. Learn more at: https://go.microsoft.com/fwlink/?linkid=2258669'
+sensitive_data_detailed_warning_message = '[Warning] This output may compromise security by showing the following secrets: {}. Learn more at: https://go.microsoft.com/fwlink/?linkid=2258669'
+
+
+class CredentialType(Enum):
+    # ([?&;]) - match character '?' or '&' or ';' as group 1, which is the prefix of signature within sas token
+    # sig= - match the literal string 'sig='
+    # [\w%-/]+ - match any word character, '-', '%', or '/' one or more times. This is the signature which needs to be redacted
+    SAS_TOKEN = (r'([?&;])sig=[\w%-/]+', r'\1sig=_REDACTED_SAS_TOKEN_SIG_', 1, 'SAS token')
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/credential_helper.py:20]: Better to explain what mean for the different levels.

File: src/azure-cli-core/azure/cli/core/commands/sensitive.py (hunk 2 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/commands/sensitive.py:34]: If we apply credential detection/redaction on _all_ commands, this `sensitive_info` flag won't be needed at all.

File: src/azure-cli-core/azure/cli/core/telemetry.py [hunk 4]
```diff
@@ -519,6 +527,14 @@ def _get_config():
     return _session.application.config
 
 
+@decorators.suppress_all_exceptions()
+def _get_secrets_warning_config():
+    show_secrets_warning = _get_config().getboolean('clients', 'show_secrets_warning', fallback=None)
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/telemetry.py:532]: `getboolean` can't take `None` as the `fallback`. The below error now appears in the `--debug` log of every command:

```
cli.azure.cli.core.decorators: Suppress exception:
Traceback (most recent call last):
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/__main__.py", line 62, in <module>
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/__main__.py", line 55, in <module>
SystemExit: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/decorators.py", line 79, in _wrapped_func
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/telemetry.py", line 532, in _get_secrets_warning_config
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\knack/config.py", line 147, in getboolean
ValueError: Not a boolean: None
```

Comment (evelyn-ys) [reply src/azure-cli-core/azure/cli/core/telemetry.py:532]: already fixed in https://github.com/Azure/azure-cli/pull/28387

**Pull request merged**